# Machine Learning

## Objectives
* Train and compare two classification models to predict fraud, and check whether the model's own reasoning agrees with the hypothesis testing done in the EDA notebook.

## Inputs
* data/cleaned/cleaned_transactions.csv

## Outputs
* outputs/tables/fraud_model.joblib - the trained model
* Evaluation charts saved to outputs/figures/

## Additional Comments
* Risk_Score is deliberately excluded from the feature set - see the leakage note in the ETL notebook.



---

# Change working directory

* We are assuming you will store the notebooks in a subfolder, therefore when running the notebook in the editor, you will need to change the working directory

We need to change the working directory from its current folder to its parent folder
* We access the current directory with os.getcwd()

In [1]:
import os
current_dir = os.getcwd()
current_dir

'/Users/apple/Desktop/Fraud_Analysis/jupyter_notebooks'

We want to make the parent of the current directory the new current directory
* os.path.dirname() gets the parent directory
* os.chir() defines the new current directory

In [2]:
os.chdir(os.path.dirname(current_dir))
print("You set a new current directory")

You set a new current directory


Confirm the new current directory

In [3]:
current_dir = os.getcwd()
current_dir

'/Users/apple/Desktop/Fraud_Analysis'

In [4]:
if os.path.basename(os.getcwd()) == 'jupyter_notebooks':
    os.chdir(os.path.dirname(os.getcwd()))

print("You set a new current directory")
print(os.getcwd())

You set a new current directory
/Users/apple/Desktop/Fraud_Analysis


## Libraries Import

In [5]:
# Libraries being used in the project
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import OneHotEncoder, StandardScaler
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import (
    classification_report, ConfusionMatrixDisplay,
    roc_auc_score, roc_curve
)
import joblib

os.makedirs('outputs/figures', exist_ok=True)
os.makedirs('outputs/tables', exist_ok=True)

## Load the Cleaned Data

The cleaned dataset from the ETL notebook is loaded here. Hour is extracted from Timestamp, the
same way it was in the EDA notebook, since it's used as a model feature below.

In [6]:
df = pd.read_csv('data/cleaned/cleaned_transactions.csv', parse_dates=['Timestamp'])
df['Hour'] = df['Timestamp'].dt.hour
df.shape

(50000, 24)

# Section 1

##  Problem Framing

This is a binary classification problem - predicting Fraud_Label. Fraud makes up 32.13% of this
dataset (established in the ETL notebook), a moderate imbalance. A model that always predicted
"not fraud" would still be right about 68% of the time without ever catching a single fraud case,
so precision, recall, and ROC-AUC are used below rather than accuracy alone.

## Excluding Risk_Score

As flagged in the ETL notebook, Risk_Score correlates with Fraud_Label at r = 0.386, because it is
itself a derived fraud-risk estimate. It is excluded from the feature set below, so the model has
to learn from the raw transaction data rather than partly relying on an existing score.

## Feature Selection and Train/Test Split

In [7]:
features = [
    'Transaction_Amount', 'Account_Balance', 'Transaction_Type', 'Device_Type',
    'Merchant_Category', 'IP_Address_Flag', 'Previous_Fraudulent_Activity',
    'Daily_Transaction_Count', 'Avg_Transaction_Amount_7d', 'Failed_Transaction_Count_7d',
    'Card_Type', 'Card_Age', 'Transaction_Distance', 'Authentication_Method',
    'Is_Weekend', 'Hour'
]
target = 'Fraud_Label'

X = df[features].copy()
y = df[target]

categorical_cols = ['Transaction_Type', 'Device_Type', 'Merchant_Category', 'Card_Type', 'Authentication_Method']
numeric_cols = [c for c in features if c not in categorical_cols]

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.25, random_state=42, stratify=y
)
print(f"Train fraud rate: {y_train.mean():.2%} | Test fraud rate: {y_test.mean():.2%}")

Train fraud rate: 32.13% | Test fraud rate: 32.14%


A stratified split is used so the train and test sets keep roughly the same fraud rate as the full
dataset - 32.13% and 32.14% respectively, confirming the split preserved the class balance.

---

# Models

## Model 1: Logistic Regression

## Model 2: Random Forest

## Comparing The Two Models

## Feature Importance: Does The Model Agree With The Hypothesis Tests?

## Saving The Model

## Conclusions

---

# Push files to Repo

* In cases where you don't need to push files to Repo, you may replace this section with "Conclusions and Next Steps" and state your conclusions and next steps.

In [ ]:
import os
try:
  # create your folder here
  # os.makedirs(name='')
except Exception as e:
  print(e)
